File: app/src/main/java/com/remotearthsolutions/expensetracker/adapters/CategoryListViewAdapter.java
package com.remotearthsolutions.expensetracker.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.adapters.viewholder.CategoryListViewHolder;
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel;

import java.util.List;

public class CategoryListViewAdapter extends RecyclerView.Adapter<CategoryListViewHolder> {

    private List<CategoryModel> categorylist;
    private CategoryListViewAdapter.OnItemClickListener listener;

    public CategoryListViewAdapter(List<CategoryModel> categorylist) {
        this.categorylist = categorylist;
    }

    @NonNull
    @Override
    public CategoryListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.custom_category_view, parent, false);
        return new CategoryListViewHolder(v, listener);
    }

    @Override
    public void onBindViewHolder(@NonNull CategoryListViewHolder holder, int position) {

        CategoryModel categoryModel = categorylist.get(position);
        holder.bind(categoryModel);
    }

    @Override
    public int getItemCount() {

        if (categorylist == null) return 0;
        return categorylist.size();
    }

    public void setOnItemClickListener(CategoryListViewAdapter.OnItemClickListener listener) {
        this.listener = listener;
    }

    public interface OnItemClickListener {
        void onItemClick(CategoryModel categoryModel);
        void onItemLongClick(CategoryModel categoryModel);
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/adapters/CategoryListViewAdapter.kt
package com.remotearthsolutions.expensetracker.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.adapters.viewholder.CategoryListViewHolder
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel

class CategoryListViewAdapter(private val categorylist: List<CategoryModel>?) :
    RecyclerView.Adapter<CategoryListViewHolder>() {
    private var listener: OnItemClickListener? = null

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): CategoryListViewHolder {
        val v = LayoutInflater.from(parent.context)
            .inflate(R.layout.custom_category_view, parent, false)
        return CategoryListViewHolder(v, listener!!)
    }

    override fun onBindViewHolder(
        holder: CategoryListViewHolder,
        position: Int
    ) {
        val categoryModel = categorylist!![position]
        holder.bind(categoryModel)
    }

    override fun getItemCount(): Int {
        return categorylist?.size ?: 0
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        this.listener = listener
    }

    interface OnItemClickListener {
        fun onItemClick(categoryModel: CategoryModel?)
    }

}